### PR TITLE
find: Implement `-newerXY`

### PR DESCRIPTION
**Pull request by hanbings**
link: https://github.com/uutils/findutils/issues/307

### CODE REVIEW

**Comment by sylvestre** [comment]
Could you please add unit and integration tests? Thanks

**Comment by hanbings** [comment]
Okay, I will add the test code as soon as possible.

**Comment by sylvestre** [comment]
Thanks

**Comment by cakebaker** [comment]
@hanbings apart from the tests, is the code ready for review?

**Comment by hanbings** [comment]
> @hanbings apart from the tests, is the code ready for review?

I'm sorry I just saw the message. Yes, I've added some test code, but there seems to be something wrong with the Github Actions being tested, and I'm not quite sure the test code meets the coverage requirements. In addition, there is another piece of file system compatibility code that is not yet ready. But yeah, ready to review. Thank you.

File: src/find/matchers/mod.rs [hunk 1]
```diff
@@ -348,6 +393,33 @@ fn build_matcher_tree(
                 i += 1;
                 Some(NewerMatcher::new(args[i])?.into_box())
             }
+            "-newerat" | "-newerBt" | "-newerct" | "-newermt" => {
```
Comment (cakebaker) [inline src/find/matchers/mod.rs:396]: This covers "only" a part of the possible combinations. From the man page:
```
-newerXY reference

Succeeds  if  timestamp  X  of the file being considered is newer than timestamp Y of the file reference.
The letters X and Y can be any of the following letters:

a   The access time of the file reference
B   The birth time of the file reference
c   The inode status change time of reference
m   The modification time of the file reference
t   reference is interpreted directly as a time

Some combinations are invalid; for example, it is invalid for X to be t. [...]
```
This means you can also have combinations like `-neweraa`, `-neweraB`, `-newerac`, `-neweram`, etc.

Comment (hanbings) [reply src/find/matchers/mod.rs:396]: Sorry, I misunderstood the documentation. I'll rewrite the code.

Comment (hanbings) [reply src/find/matchers/mod.rs]: I have completed the implementation of the parameter function described in the documentation and added a portion of the test code. Should we start the review?

File: src/find/mod.rs
Comment (cakebaker) [inline src/find/mod.rs:1]: The test function `find_newer_xy` currently tests five different things, so I suggest to break it up into five test functions to make it easier to read and understand.

Comment (hanbings) [reply src/find/mod.rs:1]: OK, I've committed the changes. Thanks.

File: src/find/mod.rs [hunk 2]
```diff
@@ -854,4 +854,112 @@ mod tests {
...
+        let args = ["-newerat", "-newerBt", "-newerct", "-newermt"];
+        let times = [""];
+
+        for (arg, time) in args.iter().zip(times.iter()) {
+            let deps = FakeDependencies::new();
+            let rc = find_main(&["find", "./test_data/simple/subdir", arg, time], &deps);
+
+            assert_eq!(rc, 1);
+        }
+    }
```
Comment (cakebaker) [inline src/find/mod.rs:964]: It's good that you test for an empty time. Unfortunately, the tested behavior differs from the behavior of GNU `find`:
```
$ find test_data/simple/subdir -newerat ""
$ echo $?
0

$ cargo run --bin find test_data/simple/subdir/ -newerat ""
Error: find: I cannot figure out how to interpret ‘’ as a date or time
$ echo $?
1
```

File: src/find/mod.rs [hunk 3]
```diff
@@ -854,4 +854,112 @@ mod tests {
...
+
+    #[test]
+    fn test_find_newer_xy_error_path() {
+        // Catch a parsing error.
+        #[cfg(target_os = "linux")]
+        let args = ["-newerat", "-newerct", "-newermt"];
+        #[cfg(not(target_os = "linux"))]
+        let args = ["-newerat", "-newerBt", "-newerct", "-newermt"];
+        let times = [""];
```
Comment (cakebaker) [inline src/find/mod.rs:956]: You don't need an array for a single value. Using `let time = "";` allows you to simplify the `for` loop below.

File: src/find/mod.rs (hunk 2 above)
Comment (hanbings) [reply src/find/mod.rs]: Excuse me, I made some modifications to the code and this feature should have been implemented. 
In addition, I added a new function test_find_newer_xy_empty_time_parameter() specifically for testing empty time string input to solve the problem of time inconsistency between the file used by the test and the local environment file (to fix submissions where CI test errors occurred earlier).

File: src/find/matchers/mod.rs [hunk 4]
```diff
@@ -252,6 +256,53 @@ fn convert_arg_to_comparable_value_and_suffix(
...
 }
 
+/// This is a function that converts a specific string format into a timestamp.
+/// It allows converting a time string of
+/// "(week abbreviation) (date), (year) (time)" to a Unix timestamp.
+/// such as: "jan 01, 2025 00:00:01" -> 1735689601000
+/// When (time) is not provided, it will be automatically filled in as 00:00:00
+/// such as: "jan 01, 2025" = "jan 01, 2025 00:00:00" -> 1735689600000
+fn parse_date_str_to_timestamps(date_str: &str) -> Option<i64> {
+    let regex_pattern = r"^(?:(\w{3} \d{2}))?(?:, (\d{4}))?(?: (\d{2}:\d{2}:\d{2}))?$";
```
Comment (sylvestre) [inline src/find/matchers/mod.rs:266]: please use this kind of patterns:
make it easier to understand 

    let regex_pattern = r#"^(?P<month_day>\w{3} \d{2})?(?:, (?P<year>\d{4}))?(?: (?P<time>\d{2}:\d{2}:\d{2}))?$"#;

File: src/find/matchers/time.rs [hunk 5]
```diff
@@ -362,4 +504,155 @@ mod tests {
...
+        let file_info = get_dir_entry_for(&temp_dir.path().to_string_lossy(), "foo");
+        assert!(
+            created_matcher.matches(&file_info, &mut deps.new_matcher_io()),
+            "file created time should after 'time'"
+        );
+
+        // accessed time test
+        let accessed_matcher = NewerTimeMatcher::new(NewerOptionType::Accessed, time);
+
+        thread::sleep(Duration::from_secs(2));
```
Comment (sylvestre) [inline src/find/matchers/time.rs:594]: why sleep ?

File: src/find/matchers/mod.rs (hunk 4 above)
Comment (hanbings) [reply src/find/matchers/mod.rs:266]: Okay, this way of writing looks clearer! I will change it.

File: src/find/matchers/time.rs (hunk 5 above)
Comment (hanbings) [reply src/find/matchers/time.rs]: This is my oversight, sleep() is not really needed here, and there are two other codes with the same problem. They have been modified in the commit. Thanks!

File: src/find/matchers/time.rs [hunk 6]
```diff
@@ -362,4 +504,149 @@ mod tests {
...
+            .now()
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .unwrap()
+            .as_millis()
+            .try_into()
+            .unwrap();
+
+        let created_matcher = NewerTimeMatcher::new(NewerOptionType::Birthed, time);
+
+        thread::sleep(Duration::from_secs(2));
```
Comment (sylvestre) [inline src/find/matchers/time.rs:575]: same, is this sleep needed?

Comment (hanbings) [reply src/find/matchers/time.rs:575]: This sleep() is used to ensure that there is a small distance between the time when assert() is executed and the time when the file was created. In fact, this code fails the test on my machine after removing the sleep().

and...
Currently the i64 type microsecond timestamp is used to determine the time. I have noticed that it is not accurate enough and I am going to change it to nanoseconds.

Comment (sylvestre) [reply src/find/matchers/time.rs:575]: two seconds is long, try 0.1 ?

Comment (hanbings) [reply src/find/matchers/time.rs]: `Duration::from_secs()` does not accept floating point arguments, maybe `Duration::from_millis(100)` would also be suitable?

**Pull request merged**